net.go

```diff
@@ -90,7 +91,43 @@ func Dial(network, addr string, ctx *Ctx, flags DialFlags) (*Conn, error) {
 func DialTimeout(network, addr string, timeout time.Duration, ctx *Ctx,
 	flags DialFlags) (*Conn, error) {
 	d := net.Dialer{Timeout: timeout}
-	return dialSession(d, network, addr, ctx, flags, nil)
+	ctx, err := prepareCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := createConnection(context.Background(), d, network, addr)
+	if err != nil {
+		return nil, err
+	}
+	client, err := createSession(conn, flags, addr, ctx, nil)
+	if err != nil {
+		conn.Close()
+	}
+	return client, err
+}
+
+// DialContext acts like Dial but takes a context for network dial.
+//
+// The context includes only network dial. It does not include OpenSSL calls.
+//
+// See func Dial for a description of the network, addr, ctx and flags
+// parameters.
+func DialContext(context context.Context, network, addr string,
+	ctx *Ctx, flags DialFlags) (*Conn, error) {
+	d := net.Dialer{}
+	ctx, err := prepareCtx(ctx)
+	if err != nil {
+		return nil, err
+	}
+	conn, err := createConnection(context, d, network, addr)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
It is very likely that a connection will fail, so we will not need a ssl context object.
```suggestion
	conn, err := createConnection(context, d, network, addr)
	if err != nil {
		return nil, err
	}
	ctx, err := prepareCtx(ctx)
	if err != nil {
		return nil, err
	}
```

Reply:
Updated.